compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -641,21 +666,15 @@ fn vtable_entries<'tcx>(
                 entries.extend(COMMON_VTABLE_ENTRIES);
             }
             VtblSegment::TraitOwnEntries { trait_ref, emit_vptr } => {
-                let trait_methods = tcx
-                    .associated_items(trait_ref.def_id())
-                    .in_definition_order()
-                    .filter(|item| item.kind == ty::AssocKind::Fn);
-                // Now list each method's DefId and InternalSubsts (for within its trait).
-                // If the method can never be called from this object, produce `Vacant`.
-                let own_entries = trait_methods.map(move |trait_method| {
-                    debug!("vtable_entries: trait_method={:?}", trait_method);
-                    let def_id = trait_method.def_id;
-
-                    // Some methods cannot be called on an object; skip those.
-                    if !is_vtable_safe_method(tcx, trait_ref.def_id(), &trait_method) {
-                        debug!("vtable_entries: not vtable safe");
-                        return VtblEntry::Vacant;
```

Comment:
Is there any reason why we cannot remove the vtable entries for the "impossible predicates" situation below and remove the `Vacant` variant entirely?

Doesn't need to be considered for this PR, of course, just a passing thought.

Reply:
```rust
trait X {
    fn foo(&self)
    where
        Self: Send,
    {
    }
}

impl X for () {
    fn foo(&self) {}
}

impl X for *mut () {}

pub fn main() {
    <dyn X + Send as X>::foo(&());
}
```

We need to generate `X::foo` for `()`, but we cannot generate `X::foo` for `*mut ()`.